File: bindings/swift/Sources/EasyDogeKM/WireCodec.swift
import Foundation

/// Raised when a JavaScript-facing value cannot be converted losslessly.
public enum WireCodecError: Error, Equatable, LocalizedError {
    case invalidKoinu(field: String)
    case invalidInteger(field: String, range: String)
    case unrepresentable(field: String)

    public var errorDescription: String? {
        switch self {
        case let .invalidKoinu(field):
            return "Invalid \(field): koinu amounts must be a canonical decimal string between 0 and 18446744073709551615"
        case let .invalidInteger(field, range):
            return "Invalid \(field): expected an integer in \(range)"
        case let .unrepresentable(field):
            return "Cannot return \(field): value exceeds JavaScript's safe integer range"
        }
    }
}

/// Lossless conversions between JavaScript-facing values and Rust integer types.
public enum WireCodec {
    public static let maxSafeInteger: UInt64 = 9_007_199_254_740_991

    public static func koinu(_ value: Any?, field: String) throws -> UInt64 {
        guard let text = value as? String, isCanonicalDecimal(text), let parsed = UInt64(text) else {
            throw WireCodecError.invalidKoinu(field: field)
        }
        return parsed
    }

    public static func optionalKoinu(_ value: Any?, field: String) throws -> UInt64? {
        if value == nil || value is NSNull { return nil }
        return try koinu(value, field: field)
    }

    public static func koinuString(_ value: UInt64) -> String {
        String(value)
    }

    public static func uint64(_ value: Any?, field: String) throws -> UInt64 {
        try integral(value, field: field, max: maxSafeInteger)
    }

    public static func uint32(_ value: Any?, field: String) throws -> UInt32 {
        UInt32(try integral(value, field: field, max: UInt64(UInt32.max)))
    }

    public static func uint16(_ value: Any?, field: String) throws -> UInt16 {
        UInt16(try integral(value, field: field, max: UInt64(UInt16.max)))
    }

    public static func uint8(_ value: Any?, field: String) throws -> UInt8 {
        UInt8(try integral(value, field: field, max: UInt64(UInt8.max)))
    }

    public static func optionalUInt8(_ value: Any?, field: String) throws -> UInt8? {
        if value == nil || value is NSNull { return nil }
        return try uint8(value, field: field)
    }

    public static func int32(_ value: Any?, field: String) throws -> Int32 {
        Int32(try integral(value, field: field, max: UInt64(Int32.max)))
    }

    public static func safeInteger(_ value: UInt64, field: String) throws -> Int {
        guard value <= maxSafeInteger else { throw WireCodecError.unrepresentable(field: field) }
        return Int(value)
    }

    private static func isCanonicalDecimal(_ text: String) -> Bool {
        guard !text.isEmpty, text.count <= 20, text.allSatisfy({ $0.isASCII && $0.isNumber }) else {
            return false
        }
        return text == "0" || !text.hasPrefix("0")
    }

    private static func integral(_ value: Any?, field: String, max: UInt64) throws -> UInt64 {
        let range = "0...\(max)"
        if value is Bool { throw WireCodecError.invalidInteger(field: field, range: range) }
        let double: Double
        switch value {
        case let int as Int:
            guard int >= 0, UInt64(int) <= max else { throw WireCodecError.invalidInteger(field: field, range: range) }
            return UInt64(int)
        case let value as Double:
            double = value
        case let number as NSNumber:
            double = number.doubleValue
        default:
            throw WireCodecError.invalidInteger(field: field, range: range)
        }
        guard double.isFinite, double >= 0, double.rounded(.towardZero) == double, double <= Double(max) else {
            throw WireCodecError.invalidInteger(field: field, range: range)
        }
        return UInt64(double)
    }
}

File: bindings/kotlin/easydoge-km/src/main/java/io/easydoge/km/WireCodec.kt
package io.easydoge.km

/** Raised when a JavaScript-facing value cannot be converted losslessly. */
class WireCodecException(message: String) : IllegalArgumentException(message)

/** Lossless conversions between JavaScript-facing values and Rust integer types. */
object WireCodec {
    const val MAX_SAFE_INTEGER: Long = 9_007_199_254_740_991L
    private val CANONICAL_DECIMAL = Regex("0|[1-9][0-9]*")

    fun koinu(value: Any?, field: String): ULong {
        val text = value as? String
        if (text == null || text.length > 20 || !CANONICAL_DECIMAL.matches(text)) {
            throw invalidKoinu(field)
        }
        return text.toULongOrNull() ?: throw invalidKoinu(field)
    }

    fun optionalKoinu(value: Any?, field: String): ULong? = if (value == null) null else koinu(value, field)

    fun koinuString(value: ULong): String = value.toString()

    fun uint64(value: Any?, field: String): ULong = integral(value, field, MAX_SAFE_INTEGER).toULong()

    fun uint32(value: Any?, field: String): UInt = integral(value, field, UInt.MAX_VALUE.toLong()).toUInt()

    fun uint16(value: Any?, field: String): UShort = integral(value, field, UShort.MAX_VALUE.toLong()).toUShort()

    fun uint8(value: Any?, field: String): UByte = integral(value, field, UByte.MAX_VALUE.toLong()).toUByte()

    fun optionalUInt8(value: Any?, field: String): UByte? = if (value == null) null else uint8(value, field)

    fun int32(value: Any?, field: String): Int = integral(value, field, Int.MAX_VALUE.toLong()).toInt()

    fun safeInteger(value: ULong, field: String): Long {
        if (value > MAX_SAFE_INTEGER.toULong()) {
            throw WireCodecException("Cannot return $field: value exceeds JavaScript's safe integer range")
        }
        return value.toLong()
    }

    private fun invalidKoinu(field: String) =
        WireCodecException("Invalid $field: koinu amounts must be a canonical decimal string between 0 and 18446744073709551615")

    private fun integral(value: Any?, field: String, max: Long): Long {
        val range = "0...$max"
        val invalid = { WireCodecException("Invalid $field: expected an integer in $range") }
        return when (value) {
            is Boolean -> throw invalid()
            is Int, is Long, is Short, is Byte -> {
                val long = (value as Number).toLong()
                if (long < 0 || long > max) throw invalid()
                long
            }
            is Number -> {
                val double = value.toDouble()
                if (!double.isFinite() || double < 0 || double != Math.floor(double) || double > max.toDouble()) {
                    throw invalid()
                }
                double.toLong()
            }
            else -> throw invalid()
        }
    }
}
